doc/designs/thread-api.md

```diff
@@ -0,0 +1,304 @@
+Thread Pool Support
+===================
+
+OpenSSL wishes to support the internal use of threads for purposes of
+concurrency and parallelism in some circumstances. There are various reasons why
+this is desirable:
+
+  - Some algorithms are designed to be run in parallel (Argon2);
+  - Some transports (e.g. QUIC, DTLS) may need to handle timer events
+    independently of application calls to OpenSSL.
+
+To support this end, OpenSSL can manage an internal thread pool. Tasks can be
+scheduled on the internal thread pool.
+
+There are three options given to an application which wants to use the thread
+pool functionality:
+
+  1. Use OpenSSL's default thread pool functionality.
+
+  2. Use OpenSSL's thread pool functionality with a custom thread spawner
+     method.
+
+  3. Provide threads manually to OpenSSL for use as part of its thread pool.
+
+Default Model
+-------------
+
+In the default model, OpenSSL creates and manages threads up to a maximum
+number of threads authorized by the application.
+
+The application enables thread pooling by calling the following function
+during its initialisation:
+
+```c
+/*
+ * Set the maximum number of threads to be used by the thread pool.
+ *
+ * If the argument is 0, thread pooling is disabled. OpenSSL will not create any
+ * threads and existing threads in the thread pool will be torn down.
```

Comment:
Is the tearing down of existing threads synchronous with this API (i.e., this API blocks until the threads terminate)?

Reply:
A good question. I'd say no since existing tasks running will need to finish.

We could have a function
```
int OSSL_stop_all_threads_sync(void);
```
which is equivalent to calling the function with 0 and then waiting for all threads to exit.